src/Raven.Server/NotificationCenter/OutOfMemoryNotifications.cs

```diff
@@ -0,0 +1,92 @@
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Raven.Server.NotificationCenter.Notifications;
+using Raven.Server.NotificationCenter.Notifications.Details;
+using Sparrow;
+using Sparrow.LowMemory;
+using Voron;
+
+namespace Raven.Server.NotificationCenter
+{
+    public class OutOfMemoryNotifications
+    {
+        private readonly TimeSpan _updateFrequency = TimeSpan.FromSeconds(15);
+        private readonly ConditionalWeakTable<StorageEnvironment, ConcurrentDictionary<Type, NotificationTime>> _notificationsMetadataTable = 
+            new ConditionalWeakTable<StorageEnvironment, ConcurrentDictionary<Type, NotificationTime>>();
+        private readonly NotificationCenter _notificationsCenter;
+
+        public OutOfMemoryNotifications(NotificationCenter notificationsCenter)
+        {
+            _notificationsCenter = notificationsCenter;
+        }
+
+        public void Add(StorageEnvironment environment, Exception exception)
+        {
+            var notificationsMetadata = _notificationsMetadataTable.GetOrCreateValue(environment);
+            if (notificationsMetadata.TryGetValue(exception.GetType(), out var notificationMetadata))
+            {
+                if (DateTime.Now - notificationMetadata.Time < _updateFrequency)
+                    return;
+
+                if (Interlocked.CompareExchange(ref notificationMetadata.IsInProgress, 1, 0) == 1)
+                    return;
+
+                notificationMetadata.Time = DateTime.Now;
+            }
+            else
+            {
+                notificationMetadata = new NotificationTime
+                {
+                    Time = DateTime.Now,
+                    IsInProgress = 1
+                };
+                if (notificationsMetadata.TryAdd(exception.GetType(), notificationMetadata) == false)
+                    return;
+
+                //We are in low of memory so we want to minimize allocations
+                notificationMetadata.Key = $"{environment}:{exception.GetType()}";
+                notificationMetadata.Title = $"Out of memory occurred for '{environment}'";
+            }
+
+            var alert = AlertRaised.Create(
+                null,
+                notificationMetadata.Title,
+                exception.Message,
+                AlertType.OutOfMemoryException,
+                NotificationSeverity.Error,
+                notificationMetadata.Key,
+                OutOfMemoryDetails(exception));
+
+            _notificationsCenter.Add(alert);
+
+            Volatile.Write(ref notificationMetadata.IsInProgress, 0);
+        }
+
+        private static MessageDetails OutOfMemoryDetails(Exception exception)
+        {
+            var memoryInfo = MemoryInformation.GetMemoryInformationUsingOneTimeSmapsReader();
+
+            return new MessageDetails
+            {
+                Message = $"Managed memory: {new Size(AbstractLowMemoryMonitor.GetManagedMemoryInBytes(), SizeUnit.Bytes)}, " +
```

Comment:
In continuation to a conversation with @haludi, created this issue: https://issues.hibernatingrhinos.com/issue/RavenDB-13851 for now, we inherit this code from previous state, so I'm fine with it.